dex/runner.go

```diff
@@ -86,38 +87,94 @@ type Connector interface {
 
 // ConnectionMaster manages a Connector.
 type ConnectionMaster struct {
-	contextManager
-	wg        *sync.WaitGroup
 	connector Connector
+	cancel    context.CancelFunc
+
+	// To permit calls to Done/On before Connect, guard the done channel.
+	mtx  sync.RWMutex
+	done chan struct{}
 }
 
-// NewConnectionMaster is the constructor for a new ConnectionMaster.
+// NewConnectionMaster creates a new ConnectionMaster. The Connect method should
+// be used before Disconnect. The On, Done, and Wait methods may be used at any
+// time. However, prior to Connect, Wait and Done immediately return and signal
+// completion, respectively.
 func NewConnectionMaster(c Connector) *ConnectionMaster {
 	return &ConnectionMaster{
 		connector: c,
 	}
 }
 
 // Connect connects the Connector, and returns any initial connection error. Use
-// Disconnect to shut down the Connector.
+// Disconnect to shut down the Connector. Connect is not safe for concurrent use
+// with Disconnect, and must be used before Disconnect.
 func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
-	c.init(ctx)
+	if c.On() { // probably a bug in the consumer
+		return errors.New("already running")
+	}
+
+	// Attempt to start the Connector.
+	ctx, cancel := context.WithCancel(ctx)
+	wg, err := c.connector.Connect(ctx)
+	if err != nil {
+		cancel() // no context leak
+		return err
+	}
+
+	// It's running, enable Disconnect.
+	c.cancel = cancel // caller should synchronize Connect/Disconnect calls
+
+	// Done and On may be checked at any time.
+	done := make(chan struct{})
 	c.mtx.Lock()
-	c.wg, err = c.connector.Connect(c.ctx)
+	c.done = done // after setting cancel
 	c.mtx.Unlock()
-	return err
+
+	go func() { // capture the local variables
+		wg.Wait()
+		cancel() // if the Connector just died on its own, don't leak the context
+		close(done)
+	}()
+
+	return nil
 }
 
-// Disconnect closes the connection and waits for shutdown.
-func (c *ConnectionMaster) Disconnect() {
-	c.cancel()
+// Done returns a channel that is closed when the Connector's WaitGroup is done.
+// If called before Connect, a closed channel is returned.
+func (c *ConnectionMaster) Done() <-chan struct{} {
 	c.mtx.RLock()
-	c.wg.Wait()
-	c.mtx.RUnlock()
+	defer c.mtx.RUnlock()
+	if c.done == nil { // before first Connect
+		done := make(chan struct{})
+		close(done)
+		return done
+	}
+	return c.done
+}
+
+// On indicates if the Connector is running. This returns false if never
+// connected, or if the Connector has completed shut down.
+func (c *ConnectionMaster) On() bool {
+	select {
+	case <-c.Done():
+		return false
```

Comment:
Although `Wait` has the same semantics as before, the `On` method now correlates to the `Wait` method's timing instead of just when the inner context is canceled.  This new way is actually a bit more consistent since depending on how the Connector died (inner context canceled vs. parent context canceled vs. internally-initiated shutdown) `On` would previously indicate slightly different stages of the `Connector`'s lifecycle (signaled to shut down vs actually shut down).